File: golang/tests/pkg/unit/key_normalizer_test.go
package unit

import (
	"testing"

	"github.com/aliyun/wuying-agentbay-sdk/golang/pkg/agentbay"
	"github.com/stretchr/testify/assert"
)

// TestNormalizeSingleKey_LetterKeys tests normalization of letter keys
func TestNormalizeSingleKey_LetterKeys(t *testing.T) {
	assert.Equal(t, "a", agentbay.NormalizeSingleKey("a"), "Lowercase letter should remain lowercase")
	assert.Equal(t, "a", agentbay.NormalizeSingleKey("A"), "Uppercase letter should be converted to lowercase")
	assert.Equal(t, "z", agentbay.NormalizeSingleKey("Z"), "Uppercase Z should be converted to lowercase z")
}

// TestNormalizeSingleKey_NumberKeys tests normalization of number keys
func TestNormalizeSingleKey_NumberKeys(t *testing.T) {
	assert.Equal(t, "0", agentbay.NormalizeSingleKey("0"), "Number key should remain as-is")
	assert.Equal(t, "1", agentbay.NormalizeSingleKey("1"), "Number key should remain as-is")
	assert.Equal(t, "9", agentbay.NormalizeSingleKey("9"), "Number key should remain as-is")
}

// TestNormalizeSingleKey_ModifierKeys tests normalization of modifier keys
func TestNormalizeSingleKey_ModifierKeys(t *testing.T) {
	// Ctrl variants
	assert.Equal(t, "Ctrl", agentbay.NormalizeSingleKey("ctrl"), "ctrl should normalize to Ctrl")
	assert.Equal(t, "Ctrl", agentbay.NormalizeSingleKey("CTRL"), "CTRL should normalize to Ctrl")
	assert.Equal(t, "Ctrl", agentbay.NormalizeSingleKey("Ctrl"), "Ctrl should remain Ctrl")
	assert.Equal(t, "Ctrl", agentbay.NormalizeSingleKey("control"), "control should normalize to Ctrl")

	// Alt variants
	assert.Equal(t, "Alt", agentbay.NormalizeSingleKey("alt"), "alt should normalize to Alt")
	assert.Equal(t, "Alt", agentbay.NormalizeSingleKey("ALT"), "ALT should normalize to Alt")
	assert.Equal(t, "Alt", agentbay.NormalizeSingleKey("Alt"), "Alt should remain Alt")

	// Shift variants
	assert.Equal(t, "Shift", agentbay.NormalizeSingleKey("shift"), "shift should normalize to Shift")
	assert.Equal(t, "Shift", agentbay.NormalizeSingleKey("SHIFT"), "SHIFT should normalize to Shift")
	assert.Equal(t, "Shift", agentbay.NormalizeSingleKey("Shift"), "Shift should remain Shift")

	// Win variants
	assert.Equal(t, "Win", agentbay.NormalizeSingleKey("win"), "win should normalize to Win")
	assert.Equal(t, "Win", agentbay.NormalizeSingleKey("WIN"), "WIN should normalize to Win")
	assert.Equal(t, "Win", agentbay.NormalizeSingleKey("windows"), "windows should normalize to Win")
	assert.Equal(t, "Win", agentbay.NormalizeSingleKey("Windows"), "Windows should normalize to Win")

	// Command variants
	assert.Equal(t, "Command", agentbay.NormalizeSingleKey("command"), "command should normalize to Command")
	assert.Equal(t, "Command", agentbay.NormalizeSingleKey("cmd"), "cmd should normalize to Command")

	// Meta/Super
	assert.Equal(t, "Meta", agentbay.NormalizeSingleKey("meta"), "meta should normalize to Meta")
	assert.Equal(t, "Super", agentbay.NormalizeSingleKey("super"), "super should normalize to Super")
}

// TestNormalizeSingleKey_FunctionKeys tests normalization of function keys
func TestNormalizeSingleKey_FunctionKeys(t *testing.T) {
	// Tab
	assert.Equal(t, "Tab", agentbay.NormalizeSingleKey("tab"), "tab should normalize to Tab")
	assert.Equal(t, "Tab", agentbay.NormalizeSingleKey("TAB"), "TAB should normalize to Tab")

	// Enter variants
	assert.Equal(t, "Enter", agentbay.NormalizeSingleKey("enter"), "enter should normalize to Enter")
	assert.Equal(t, "Enter", agentbay.NormalizeSingleKey("ENTER"), "ENTER should normalize to Enter")
	assert.Equal(t, "Enter", agentbay.NormalizeSingleKey("return"), "return should normalize to Enter")

	// Escape variants
	assert.Equal(t, "Escape", agentbay.NormalizeSingleKey("escape"), "escape should normalize to Escape")
	assert.Equal(t, "Escape", agentbay.NormalizeSingleKey("ESC"), "ESC should normalize to Escape")
	assert.Equal(t, "Escape", agentbay.NormalizeSingleKey("esc"), "esc should normalize to Escape")

	// Delete variants
	assert.Equal(t, "Delete", agentbay.NormalizeSingleKey("delete"), "delete should normalize to Delete")
	assert.Equal(t, "Delete", agentbay.NormalizeSingleKey("del"), "del should normalize to Delete")

	// Other function keys
	assert.Equal(t, "Backspace", agentbay.NormalizeSingleKey("backspace"), "backspace should normalize to Backspace")
	assert.Equal(t, "Insert", agentbay.NormalizeSingleKey("insert"), "insert should normalize to Insert")
	assert.Equal(t, "Insert", agentbay.NormalizeSingleKey("ins"), "ins should normalize to Insert")
	assert.Equal(t, "Home", agentbay.NormalizeSingleKey("home"), "home should normalize to Home")
	assert.Equal(t, "End", agentbay.NormalizeSingleKey("end"), "end should normalize to End")
	assert.Equal(t, "PageUp", agentbay.NormalizeSingleKey("pageup"), "pageup should normalize to PageUp")
	assert.Equal(t, "PageDown", agentbay.NormalizeSingleKey("pagedown"), "pagedown should normalize to PageDown")
	assert.Equal(t, "Space", agentbay.NormalizeSingleKey("space"), "space should normalize to Space")
}

// TestNormalizeSingleKey_FKeys tests normalization of F-keys
func TestNormalizeSingleKey_FKeys(t *testing.T) {
	assert.Equal(t, "F1", agentbay.NormalizeSingleKey("f1"), "f1 should normalize to F1")
	assert.Equal(t, "F1", agentbay.NormalizeSingleKey("F1"), "F1 should remain F1")
	assert.Equal(t, "F12", agentbay.NormalizeSingleKey("f12"), "f12 should normalize to F12")
	assert.Equal(t, "F24", agentbay.NormalizeSingleKey("f24"), "f24 should normalize to F24")
}

// TestNormalizeSingleKey_ArrowKeys tests normalization of arrow keys
func TestNormalizeSingleKey_ArrowKeys(t *testing.T) {
	// Full names
	assert.Equal(t, "ArrowUp", agentbay.NormalizeSingleKey("arrowup"), "arrowup should normalize to ArrowUp")
	assert.Equal(t, "ArrowUp", agentbay.NormalizeSingleKey("ARROWUP"), "ARROWUP should normalize to ArrowUp")
	assert.Equal(t, "ArrowDown", agentbay.NormalizeSingleKey("arrowdown"), "arrowdown should normalize to ArrowDown")
	assert.Equal(t, "ArrowLeft", agentbay.NormalizeSingleKey("arrowleft"), "arrowleft should normalize to ArrowLeft")
	assert.Equal(t, "ArrowRight", agentbay.NormalizeSingleKey("arrowright"), "arrowright should normalize to ArrowRight")

	// Short names
	assert.Equal(t, "ArrowUp", agentbay.NormalizeSingleKey("up"), "up should normalize to ArrowUp")
	assert.Equal(t, "ArrowDown", agentbay.NormalizeSingleKey("down"), "down should normalize to ArrowDown")
	assert.Equal(t, "ArrowLeft", agentbay.NormalizeSingleKey("left"), "left should normalize to ArrowLeft")
	assert.Equal(t, "ArrowRight", agentbay.NormalizeSingleKey("right"), "right should normalize to ArrowRight")
}

// TestNormalizeKeys_SingleLetter tests normalization of single letter keys
func TestNormalizeKeys_SingleLetter(t *testing.T) {
	result := agentbay.NormalizeKeys([]string{"a"})
	assert.Equal(t, []string{"a"}, result, "Lowercase letter should remain lowercase")

	result = agentbay.NormalizeKeys([]string{"A"})
	assert.Equal(t, []string{"a"}, result, "Uppercase letter should be converted to lowercase")
}

// TestNormalizeKeys_ModifierCombinations tests normalization of modifier key combinations
func TestNormalizeKeys_ModifierCombinations(t *testing.T) {
	// Ctrl+C (lowercase)
	result := agentbay.NormalizeKeys([]string{"ctrl", "c"})
	assert.Equal(t, []string{"Ctrl", "c"}, result, "ctrl+c should normalize to Ctrl+c")

	// CTRL+C (uppercase)
	result = agentbay.NormalizeKeys([]string{"CTRL", "C"})
	assert.Equal(t, []string{"Ctrl", "c"}, result, "CTRL+C should normalize to Ctrl+c")

	// Ctrl+Shift+Z
	result = agentbay.NormalizeKeys([]string{"ctrl", "shift", "z"})
	assert.Equal(t, []string{"Ctrl", "Shift", "z"}, result, "ctrl+shift+z should normalize correctly")

	// Alt+Tab
	result = agentbay.NormalizeKeys([]string{"alt", "tab"})
	assert.Equal(t, []string{"Alt", "Tab"}, result, "alt+tab should normalize to Alt+Tab")

	// Win+D
	result = agentbay.NormalizeKeys([]string{"win", "d"})
	assert.Equal(t, []string{"Win", "d"}, result, "win+d should normalize to Win+d")
}

// TestNormalizeKeys_MixedCase tests normalization of mixed case combinations
func TestNormalizeKeys_MixedCase(t *testing.T) {
	result := agentbay.NormalizeKeys([]string{"CTRL", "a"})
	assert.Equal(t, []string{"Ctrl", "a"}, result, "CTRL+a should normalize to Ctrl+a")

	result = agentbay.NormalizeKeys([]string{"alt", "TAB"})
	assert.Equal(t, []string{"Alt", "Tab"}, result, "alt+TAB should normalize to Alt+Tab")

	result = agentbay.NormalizeKeys([]string{"ALT", "tab"})
	assert.Equal(t, []string{"Alt", "Tab"}, result, "ALT+tab should normalize to Alt+Tab")
}

// TestNormalizeKeys_FKeysCombinations tests normalization of F-key combinations
func TestNormalizeKeys_FKeysCombinations(t *testing.T) {
	result := agentbay.NormalizeKeys([]string{"alt", "f4"})
	assert.Equal(t, []string{"Alt", "F4"}, result, "alt+f4 should normalize to Alt+F4")

	result = agentbay.NormalizeKeys([]string{"ALT", "F4"})
	assert.Equal(t, []string{"Alt", "F4"}, result, "ALT+F4 should normalize to Alt+F4")
}

// TestNormalizeKeys_ArrowKeysCombinations tests normalization of arrow key combinations
func TestNormalizeKeys_ArrowKeysCombinations(t *testing.T) {
	result := agentbay.NormalizeKeys([]string{"shift", "arrowleft"})
	assert.Equal(t, []string{"Shift", "ArrowLeft"}, result, "shift+arrowleft should normalize correctly")

	result = agentbay.NormalizeKeys([]string{"SHIFT", "LEFT"})
	assert.Equal(t, []string{"Shift", "ArrowLeft"}, result, "SHIFT+LEFT should normalize correctly")
}

// TestNormalizeKeys_Aliases tests normalization of key aliases
func TestNormalizeKeys_Aliases(t *testing.T) {
	// control -> Ctrl
	result := agentbay.NormalizeKeys([]string{"control", "c"})
	assert.Equal(t, []string{"Ctrl", "c"}, result, "control should normalize to Ctrl")

	// windows -> Win
	result = agentbay.NormalizeKeys([]string{"windows", "e"})
	assert.Equal(t, []string{"Win", "e"}, result, "windows should normalize to Win")

	// esc -> Escape
	result = agentbay.NormalizeKeys([]string{"esc"})
	assert.Equal(t, []string{"Escape"}, result, "esc should normalize to Escape")

	// return -> Enter
	result = agentbay.NormalizeKeys([]string{"return"})
	assert.Equal(t, []string{"Enter"}, result, "return should normalize to Enter")

	// del -> Delete
	result = agentbay.NormalizeKeys([]string{"del"})
	assert.Equal(t, []string{"Delete"}, result, "del should normalize to Delete")
}

// TestNormalizeKeys_EmptyInput tests normalization of empty input
func TestNormalizeKeys_EmptyInput(t *testing.T) {
	result := agentbay.NormalizeKeys([]string{})
	assert.Equal(t, []string{}, result, "Empty array should return empty array")

	result = agentbay.NormalizeKeys(nil)
	assert.Nil(t, result, "Nil input should return nil")
}

// TestNormalizeKeys_RealWorldUseCases tests real-world use cases
func TestNormalizeKeys_RealWorldUseCases(t *testing.T) {
	// Copy (Ctrl+C)
	result := agentbay.NormalizeKeys([]string{"ctrl", "c"})
	assert.Equal(t, []string{"Ctrl", "c"}, result, "Copy shortcut should normalize correctly")

	// Paste (Ctrl+V)
	result = agentbay.NormalizeKeys([]string{"CTRL", "V"})
	assert.Equal(t, []string{"Ctrl", "v"}, result, "Paste shortcut should normalize correctly")

	// Undo (Ctrl+Z)
	result = agentbay.NormalizeKeys([]string{"ctrl", "z"})
	assert.Equal(t, []string{"Ctrl", "z"}, result, "Undo shortcut should normalize correctly")

	// Redo (Ctrl+Shift+Z)
	result = agentbay.NormalizeKeys([]string{"ctrl", "shift", "z"})
	assert.Equal(t, []string{"Ctrl", "Shift", "z"}, result, "Redo shortcut should normalize correctly")

	// Alt+Tab (window switching)
	result = agentbay.NormalizeKeys([]string{"alt", "tab"})
	assert.Equal(t, []string{"Alt", "Tab"}, result, "Alt+Tab should normalize correctly")

	// Win+D (show desktop)
	result = agentbay.NormalizeKeys([]string{"win", "d"})
	assert.Equal(t, []string{"Win", "d"}, result, "Win+D should normalize correctly")

	// Alt+F4 (close window)
	result = agentbay.NormalizeKeys([]string{"alt", "f4"})
	assert.Equal(t, []string{"Alt", "F4"}, result, "Alt+F4 should normalize correctly")

	// Shift+ArrowLeft (select text)
	result = agentbay.NormalizeKeys([]string{"shift", "arrowleft"})
	assert.Equal(t, []string{"Shift", "ArrowLeft"}, result, "Shift+ArrowLeft should normalize correctly")
}

// TestGetSupportedKeyNames tests the GetSupportedKeyNames function
func TestGetSupportedKeyNames(t *testing.T) {
	mappings := agentbay.GetSupportedKeyNames()

	// Check that all expected categories exist
	assert.NotNil(t, mappings["modifierKeys"], "modifierKeys should exist")
	assert.NotNil(t, mappings["functionKeys"], "functionKeys should exist")
	assert.NotNil(t, mappings["arrowKeys"], "arrowKeys should exist")
	assert.NotNil(t, mappings["fKeys"], "fKeys should exist")
	assert.NotNil(t, mappings["specialChars"], "specialChars should exist")

	// Check some specific mappings
	assert.Equal(t, "Ctrl", mappings["modifierKeys"]["ctrl"], "ctrl should map to Ctrl")
	assert.Equal(t, "Tab", mappings["functionKeys"]["tab"], "tab should map to Tab")
	assert.Equal(t, "ArrowUp", mappings["arrowKeys"]["up"], "up should map to ArrowUp")
	assert.Equal(t, "F1", mappings["fKeys"]["f1"], "f1 should map to F1")
}

File: typescript/tests/unit/key-normalizer.test.ts
import { describe, test, expect } from "@jest/globals";
import { normalizeKeys, getSupportedKeyNames } from "../../src/key-normalizer";

describe("NormalizeSingleKey", () => {
  describe("Letter Keys", () => {
    test("should keep lowercase letters lowercase", () => {
      expect(normalizeKeys(["a"])[0]).toBe("a");
      expect(normalizeKeys(["z"])[0]).toBe("z");
    });

    test("should convert uppercase letters to lowercase", () => {
      expect(normalizeKeys(["A"])[0]).toBe("a");
      expect(normalizeKeys(["Z"])[0]).toBe("z");
    });
  });

  describe("Number Keys", () => {
    test("should keep number keys as-is", () => {
      expect(normalizeKeys(["0"])[0]).toBe("0");
      expect(normalizeKeys(["1"])[0]).toBe("1");
      expect(normalizeKeys(["9"])[0]).toBe("9");
    });
  });

  describe("Modifier Keys", () => {
    test("should normalize ctrl variants", () => {
      expect(normalizeKeys(["ctrl"])[0]).toBe("Ctrl");
      expect(normalizeKeys(["CTRL"])[0]).toBe("Ctrl");
      expect(normalizeKeys(["Ctrl"])[0]).toBe("Ctrl");
      expect(normalizeKeys(["control"])[0]).toBe("Ctrl");
    });

    test("should normalize alt variants", () => {
      expect(normalizeKeys(["alt"])[0]).toBe("Alt");
      expect(normalizeKeys(["ALT"])[0]).toBe("Alt");
      expect(normalizeKeys(["Alt"])[0]).toBe("Alt");
    });

    test("should normalize shift variants", () => {
      expect(normalizeKeys(["shift"])[0]).toBe("Shift");
      expect(normalizeKeys(["SHIFT"])[0]).toBe("Shift");
      expect(normalizeKeys(["Shift"])[0]).toBe("Shift");
    });

    test("should normalize win/windows variants", () => {
      expect(normalizeKeys(["win"])[0]).toBe("Win");
      expect(normalizeKeys(["WIN"])[0]).toBe("Win");
      expect(normalizeKeys(["windows"])[0]).toBe("Win");
      expect(normalizeKeys(["Windows"])[0]).toBe("Win");
    });

    test("should normalize command variants", () => {
      expect(normalizeKeys(["command"])[0]).toBe("Command");
      expect(normalizeKeys(["cmd"])[0]).toBe("Command");
    });

    test("should normalize meta and super", () => {
      expect(normalizeKeys(["meta"])[0]).toBe("Meta");
      expect(normalizeKeys(["super"])[0]).toBe("Super");
    });
  });

  describe("Function Keys", () => {
    test("should normalize tab variants", () => {
      expect(normalizeKeys(["tab"])[0]).toBe("Tab");
      expect(normalizeKeys(["TAB"])[0]).toBe("Tab");
      expect(normalizeKeys(["Tab"])[0]).toBe("Tab");
    });

    test("should normalize enter variants", () => {
      expect(normalizeKeys(["enter"])[0]).toBe("Enter");
      expect(normalizeKeys(["ENTER"])[0]).toBe("Enter");
      expect(normalizeKeys(["return"])[0]).toBe("Enter");
    });

    test("should normalize escape variants", () => {
      expect(normalizeKeys(["escape"])[0]).toBe("Escape");
      expect(normalizeKeys(["ESC"])[0]).toBe("Escape");
      expect(normalizeKeys(["esc"])[0]).toBe("Escape");
    });

    test("should normalize delete variants", () => {
      expect(normalizeKeys(["delete"])[0]).toBe("Delete");
      expect(normalizeKeys(["del"])[0]).toBe("Delete");
    });

    test("should normalize other function keys", () => {
      expect(normalizeKeys(["backspace"])[0]).toBe("Backspace");
      expect(normalizeKeys(["insert"])[0]).toBe("Insert");
      expect(normalizeKeys(["ins"])[0]).toBe("Insert");
      expect(normalizeKeys(["home"])[0]).toBe("Home");
      expect(normalizeKeys(["end"])[0]).toBe("End");
      expect(normalizeKeys(["pageup"])[0]).toBe("PageUp");
      expect(normalizeKeys(["pagedown"])[0]).toBe("PageDown");
      expect(normalizeKeys(["space"])[0]).toBe("Space");
    });
  });

  describe("F-Keys", () => {
    test("should normalize F-keys", () => {
      expect(normalizeKeys(["f1"])[0]).toBe("F1");
      expect(normalizeKeys(["F1"])[0]).toBe("F1");
      expect(normalizeKeys(["f12"])[0]).toBe("F12");
      expect(normalizeKeys(["f24"])[0]).toBe("F24");
    });
  });

  describe("Arrow Keys", () => {
    test("should normalize arrow keys with full names", () => {
      expect(normalizeKeys(["arrowup"])[0]).toBe("ArrowUp");
      expect(normalizeKeys(["ARROWUP"])[0]).toBe("ArrowUp");
      expect(normalizeKeys(["arrowdown"])[0]).toBe("ArrowDown");
      expect(normalizeKeys(["arrowleft"])[0]).toBe("ArrowLeft");
      expect(normalizeKeys(["arrowright"])[0]).toBe("ArrowRight");
    });

    test("should normalize arrow keys with short names", () => {
      expect(normalizeKeys(["up"])[0]).toBe("ArrowUp");
      expect(normalizeKeys(["down"])[0]).toBe("ArrowDown");
      expect(normalizeKeys(["left"])[0]).toBe("ArrowLeft");
      expect(normalizeKeys(["right"])[0]).toBe("ArrowRight");
    });
  });
});

describe("NormalizeKeys", () => {
  describe("Empty Input", () => {
    test("should handle empty array", () => {
      expect(normalizeKeys([])).toEqual([]);
    });

    test("should handle null input", () => {
      expect(normalizeKeys(null as any)).toBe(null);
    });
  });

  describe("Single Key", () => {
    test("should normalize single key", () => {
      expect(normalizeKeys(["a"])).toEqual(["a"]);
      expect(normalizeKeys(["A"])).toEqual(["a"]);
    });
  });

  describe("Modifier Combinations", () => {
    test("should normalize ctrl+c (lowercase)", () => {
      expect(normalizeKeys(["ctrl", "c"])).toEqual(["Ctrl", "c"]);
    });

    test("should normalize CTRL+C (uppercase)", () => {
      expect(normalizeKeys(["CTRL", "C"])).toEqual(["Ctrl", "c"]);
    });

    test("should normalize ctrl+shift+z", () => {
      expect(normalizeKeys(["ctrl", "shift", "z"])).toEqual(["Ctrl", "Shift", "z"]);
    });

    test("should normalize alt+tab", () => {
      expect(normalizeKeys(["alt", "tab"])).toEqual(["Alt", "Tab"]);
    });

    test("should normalize win+d", () => {
      expect(normalizeKeys(["win", "d"])).toEqual(["Win", "d"]);
    });
  });

  describe("Mixed Case", () => {
    test("should normalize CTRL+a", () => {
      expect(normalizeKeys(["CTRL", "a"])).toEqual(["Ctrl", "a"]);
    });

    test("should normalize alt+TAB", () => {
      expect(normalizeKeys(["alt", "TAB"])).toEqual(["Alt", "Tab"]);
    });

    test("should normalize ALT+tab", () => {
      expect(normalizeKeys(["ALT", "tab"])).toEqual(["Alt", "Tab"]);
    });
  });

  describe("F-Keys Combinations", () => {
    test("should normalize alt+f4", () => {
      expect(normalizeKeys(["alt", "f4"])).toEqual(["Alt", "F4"]);
    });

    test("should normalize ALT+F4", () => {
      expect(normalizeKeys(["ALT", "F4"])).toEqual(["Alt", "F4"]);
    });
  });

  describe("Arrow Keys Combinations", () => {
    test("should normalize shift+arrowleft", () => {
      expect(normalizeKeys(["shift", "arrowleft"])).toEqual(["Shift", "ArrowLeft"]);
    });

    test("should normalize SHIFT+LEFT", () => {
      expect(normalizeKeys(["SHIFT", "LEFT"])).toEqual(["Shift", "ArrowLeft"]);
    });
  });

  describe("Key Aliases", () => {
    test("should normalize control to Ctrl", () => {
      expect(normalizeKeys(["control", "c"])).toEqual(["Ctrl", "c"]);
    });

    test("should normalize windows to Win", () => {
      expect(normalizeKeys(["windows", "e"])).toEqual(["Win", "e"]);
    });

    test("should normalize esc to Escape", () => {
      expect(normalizeKeys(["esc"])).toEqual(["Escape"]);
    });

    test("should normalize return to Enter", () => {
      expect(normalizeKeys(["return"])).toEqual(["Enter"]);
    });

    test("should normalize del to Delete", () => {
      expect(normalizeKeys(["del"])).toEqual(["Delete"]);
    });
  });

  describe("Real-World Use Cases", () => {
    test("should normalize common shortcuts", () => {
      // Copy (Ctrl+C)
      expect(normalizeKeys(["ctrl", "c"])).toEqual(["Ctrl", "c"]);

      // Paste (Ctrl+V)
      expect(normalizeKeys(["CTRL", "V"])).toEqual(["Ctrl", "v"]);

      // Undo (Ctrl+Z)
      expect(normalizeKeys(["ctrl", "z"])).toEqual(["Ctrl", "z"]);

      // Redo (Ctrl+Shift+Z)
      expect(normalizeKeys(["ctrl", "shift", "z"])).toEqual(["Ctrl", "Shift", "z"]);
    });

    test("should normalize window management shortcuts", () => {
      // Alt+Tab (window switching)
      expect(normalizeKeys(["alt", "tab"])).toEqual(["Alt", "Tab"]);

      // Win+D (show desktop)
      expect(normalizeKeys(["win", "d"])).toEqual(["Win", "d"]);

      // Alt+F4 (close window)
      expect(normalizeKeys(["alt", "f4"])).toEqual(["Alt", "F4"]);
    });

    test("should normalize text editing shortcuts", () => {
      // Shift+ArrowLeft (select text)
      expect(normalizeKeys(["shift", "arrowleft"])).toEqual(["Shift", "ArrowLeft"]);

      // Ctrl+A (select all)
      expect(normalizeKeys(["ctrl", "a"])).toEqual(["Ctrl", "a"]);

      // Ctrl+Home (go to beginning)
      expect(normalizeKeys(["ctrl", "home"])).toEqual(["Ctrl", "Home"]);
    });
  });
});

describe("GetSupportedKeyNames", () => {
  test("should return object with all categories", () => {
    const mappings = getSupportedKeyNames();

    expect(mappings).toHaveProperty("modifierKeys");
    expect(mappings).toHaveProperty("functionKeys");
    expect(mappings).toHaveProperty("arrowKeys");
    expect(mappings).toHaveProperty("fKeys");
    expect(mappings).toHaveProperty("specialChars");
  });

  test("should contain correct modifier key mappings", () => {
    const mappings = getSupportedKeyNames();

    expect(mappings.modifierKeys["ctrl"]).toBe("Ctrl");
    expect(mappings.modifierKeys["alt"]).toBe("Alt");
    expect(mappings.modifierKeys["shift"]).toBe("Shift");
  });

  test("should contain correct function key mappings", () => {
    const mappings = getSupportedKeyNames();

    expect(mappings.functionKeys["tab"]).toBe("Tab");
    expect(mappings.functionKeys["enter"]).toBe("Enter");
    expect(mappings.functionKeys["escape"]).toBe("Escape");
  });

  test("should contain correct arrow key mappings", () => {
    const mappings = getSupportedKeyNames();

    expect(mappings.arrowKeys["up"]).toBe("ArrowUp");
    expect(mappings.arrowKeys["down"]).toBe("ArrowDown");
    expect(mappings.arrowKeys["left"]).toBe("ArrowLeft");
    expect(mappings.arrowKeys["right"]).toBe("ArrowRight");
  });

  test("should contain correct F-key mappings", () => {
    const mappings = getSupportedKeyNames();

    expect(mappings.fKeys["f1"]).toBe("F1");
    expect(mappings.fKeys["f12"]).toBe("F12");
    expect(mappings.fKeys["f24"]).toBe("F24");
  });
});
